### PR TITLE
[Model Compression] Add bank pruning for level pruner

### PR DESCRIPTION
**Pull request by linbinskn**
### Description ###
This pr aims to support level pruner with bank feature. Bank pruning means finegrained pruning will be done in granularity of bank. For instance, if user set number of values in per bank to 4 and sparsity to 50%, level pruner will generate the same sparse pattern with [NVIDIA Sparse TensorCore](https://developer.nvidia.com/blog/exploiting-ampere-structured-sparsity-with-cusparselt/).

### Checklist ###
  - [x] test case
  - [x] doc

### CODE REVIEW

**Comment by linbinskn** [comment]
@J-shang @littlelittlehorse
Please help review this pr, thanks.

**Comment by J-shang** [comment]
It is a good feature and I think maybe it could be more powerful. `SparsityAllocator` always corresponds to a `mode`, maybe we can add a `bank` mode here, and bank maybe can be any block size, not only one dimension, https://nni.readthedocs.io/en/stable/Compression/v2_pruning_algo.html#nni.algorithms.compression.v2.pytorch.pruning.TaylorFOWeightPruner

**Comment by linbinskn** [comment]
> It is a good feature and I think maybe it could be more powerful. `SparsityAllocator` always corresponds to a `mode`, maybe we can add a `bank` mode here, and bank maybe can be any block size, not only one dimension, https://nni.readthedocs.io/en/stable/Compression/v2_pruning_algo.html#nni.algorithms.compression.v2.pytorch.pruning.TaylorFOWeightPruner

Have updated, please help review again：）

**Comment by J-shang** [comment]
please fix ut, better to use a sparsity range

**Comment by linbinskn** [comment]
> please fix ut, better to use a sparsity range

Have fixed.
I think keeping using specific sparsity value instead of range here is better since it can help validate correctness of `balance` mode implementation (balance_gran = [5], sparsity = 0.7 -> remain 3 value per block, finally 0.6 sparsity).

**Comment by J-shang** [comment]
> > please fix ut, better to use a sparsity range
> 
> Have fixed. I think keeping using specific sparsity value instead of range here is better since it can help validate correctness of `balance` mode implementation (balance_gran = [5], sparsity = 0.7 -> remain 3 value per block, finally 0.6 sparsity).

Got it

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 1]
```diff
@@ -137,9 +138,51 @@ class LevelPruner(BasicPruner):
...
+                sparsity: 25%
+                balance_gran: [32, 32]
+
+                pruning result: Weight tensor whose shape is [64, 64] will be split into 4 [32, 32] sub blocks.
+                                Each sub block will be pruned 256 values.
+                
     """
 
-    def __init__(self, model: Module, config_list: List[Dict]):
+    def __init__(self, model: Module, config_list: List[Dict], mode: str = "normal", balance_gran: list = None):
```
Comment (J-shang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:183]: `balance_gran: Optional[List] = None`

File: nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py [hunk 2]
```diff
@@ -40,6 +41,60 @@ def generate_sparsity(self, metrics: Dict[str, Tensor]) -> Dict[str, Dict[str, T
...
+                sub_mask_str =  "mask[{}] = mask_bank".format(index_str)
+                metric_bank = eval(sub_metric_str)
+                prune_num = int(sparsity_rate * metric_bank.numel())
+                if prune_num == 0:
+                    threshold = metric_bank.min() -1
+                else:
+                    threshold = torch.topk(metric_bank.reshape(-1), prune_num, largest=False)[0].max()
+                # mask_bank will be used in exec(sub_mask_str)
+                mask_bank = torch.gt(metric_bank, threshold).type_as(metric_bank)
+                exec(sub_mask_str)
```
Comment (J-shang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py:92]: Amazing loop! 😆

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 3]
```diff
@@ -137,9 +138,51 @@ class LevelPruner(BasicPruner):
...
+
+            example 1:
+                operation: Linear
+                pruning tensor: weight
+                pruning tensor shape: [32, 32]
+                sparsity: 50%
+                balance_gran: [4]
+
+                pruning result: Weight tensor whose shape is [32, 32] will be split into 256 [1, 4] sub blocks.
+                                Each sub block will be pruned 2 values.
```
Comment (J-shang) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:169]: we should note we will align from the back.

**Review by J-shang — APPROVED** [review]

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 1 above)
Comment (linbinskn) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py]: Have modified.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 3 above)
Comment (linbinskn) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:173]: Good point, added.

**Review by littlelittlehorse — APPROVED** [review]

**Pull request merged**